### PR TITLE
don't change the currently logged in user

### PR DESCRIPTION
**Pull request by alexcb**
if there's a valid auth.jwt, earthly should not re-login (potentially with different credentials).

### CODE REVIEW

File: cmd/earthly/account_cmds.go [hunk 1]
```diff
@@ -535,6 +535,16 @@ func (app *earthlyApp) actionAccountLogin(cliCtx *cli.Context) error {
 		return err
 	}
 
+	loggedInEmail, authType, _, err := cloudClient.WhoAmI(cliCtx.Context)
+	if err == nil {
+		// already logged in, don't re-attempt a login
+		app.console.Printf("Logged in as %q using %s auth\n", loggedInEmail, authType)
```
Comment (alexcb) [inline cmd/earthly/account_cmds.go:541]: this is the most significant logic-related change.

**Comment by alexcb** [comment]
fixes #3099

File: scripts/tests/secrets-integration.sh [hunk 2]
```diff
@@ -58,7 +59,7 @@ echo testing that the ssh-agent only contains a single key
 test "$(ssh-add -l | wc -l)" = "1"
 
 echo "testing earthly account login works (and is using the earthly-manitou account)"
-"$earthly" account login 2>&1 | acbgrep 'other-service+earthly-manitou@earthly.dev'
+"$earthly" account login 2>&1 | acbgrep 'Logged in as "other-service+earthly@earthly.dev"'
```
Comment (alexcb) [inline scripts/tests/secrets-integration.sh:62]: whoops.... I think this is wrong. this shouldn't have changed :facepalm:

File: scripts/tests/secrets-integration.sh [hunk 3]
```diff
@@ -40,6 +41,7 @@ ID_RSA=$("$earthly" secrets --org earthly-technologies --project core get -n sec
 
 # now that we grabbed the manitou credentials, unset our token, to ensure that we're only testing using manitou's credentials
 unset EARTHLY_TOKEN
+"$earthly" account logout
```
Comment (alexcb) [inline scripts/tests/secrets-integration.sh:44]: this is now needed to force the cached jwt to be discarded.

File: cloud/account.go [hunk 4]
```diff
@@ -16,6 +16,15 @@ import (
 	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
+type AuthMethod string
+
+const (
+	SSHAuthMethod       AuthMethod = "ssh"
+	PasswordAuthMethod  AuthMethod = "password"
+	TokenAuthMethod     AuthMethod = "token"
+	CachedJWTAuthMethod AuthMethod = "cached jwt"
```
Comment (nelsam) [inline cloud/account.go:25]: super minor nit: I believe that idiomatic go puts the type first, e.g. [http methods](https://pkg.go.dev/net/http#pkg-constants). `AuthMethodSSH` and such are probably more idiomatic.

File: cloud/auth.go [hunk 5]
```diff
@@ -23,23 +23,32 @@ import (
...
 	case c.email != "" && c.password != "":
+		authMethod = "password"
 		err = c.loginWithPassword(ctx)
 	case c.authCredToken != "":
+		authMethod = "token"
 		err = c.loginWithToken(ctx)
 	default:
+		authMethod = "ssh"
 		err = c.loginWithSSH(ctx)
 	}
```
Comment (nelsam) [inline cloud/auth.go:39]: It feels like this is a good opportunity to move this to some form of `authMethod` function/method, e.g.

```go
func (c *Client) authMethod(ctx context.Context) (AuthMethod, error) {
    if c.email != "" && c.password != "" {
        return AuthMethodPassword, c.loginWithPassword(ctx)
    }
    if c.authCredToken != "" {
        return AuthMethodToken, c.loginWithToken(ctx)
    }
    return AuthMethodSSH, c.loginWithSSH(ctx)
}
```

**Review by nelsam — APPROVED** [review]
Minor style nits aside, lgtm

File: cloud/auth.go (hunk 5 above)
Comment (alexcb) [reply cloud/auth.go:39]: good idea!

Comment (alexcb) [reply cloud/auth.go:39]: I also totally missed that `authMethod = "password"` wasn't changed to use the consts.

**Pull request merged**